File: tests/Tests.FSharp/TriBoolean/CrossVerifyTests.fs
module Zeta.Tests.FSharp.TriBoolean.CrossVerifyTests

open System
open System.IO
open System.Text
open System.Text.Json
open Xunit
open Zeta.Core.FSharp.TriBoolean
open Zeta.Core.FSharp.TriBoolean.TriBoolean
open Zeta.Core.FSharp.Yaml.Dom

// ---------------------------------------------------------------------------
// Repo-root walk (Zeta.sln sentinel)
// ---------------------------------------------------------------------------

let private repoRoot () : string =
    let assembly = typeof<YamlValue>.Assembly
    let mutable dir = DirectoryInfo(Path.GetDirectoryName(assembly.Location))
    while not (isNull dir) && not (File.Exists(Path.Join(dir.FullName, "Zeta.sln"))) do
        dir <- dir.Parent
    if isNull dir then
        raise (InvalidOperationException("Could not locate repo root (Zeta.sln) from test assembly location."))
    dir.FullName

// ---------------------------------------------------------------------------
// YamlValue navigation helpers
// ---------------------------------------------------------------------------

let private mapEntries (v: YamlValue) (ctx: string) : (string * YamlValue) list =
    match v with
    | VMap entries -> entries
    | other -> raise (InvalidOperationException(sprintf "expected Map at %s, got %A" ctx other))

let private field (entries: (string * YamlValue) list) (key: string) (ctx: string) : YamlValue =
    match entries |> List.tryFind (fun (k, _) -> String.Equals(k, key, StringComparison.Ordinal)) with
    | Some(_, value) -> value
    | None -> raise (InvalidOperationException(sprintf "missing field '%s' at %s" key ctx))

let private asStr (v: YamlValue) (ctx: string) : string =
    match v with
    | VStr s -> s
    | other -> raise (InvalidOperationException(sprintf "expected Str at %s, got %A" ctx other))

// ---------------------------------------------------------------------------
// Mapping functions
// ---------------------------------------------------------------------------

let private toTri (s: string) : Tri =
    match s with
    | "T" -> Tri.T
    | "F" -> Tri.F
    | "N" -> Tri.N
    | other -> raise (InvalidOperationException(sprintf "unknown tri state: %s" other))

let private toStr (t: Tri) : string =
    match t with
    | Tri.T -> "T"
    | Tri.F -> "F"
    | Tri.N -> "N"

let private tryField (entries: (string * YamlValue) list) (key: string) : YamlValue option =
    entries |> List.tryFind (fun (k, _) -> String.Equals(k, key, StringComparison.Ordinal)) |> Option.map snd

let private asNumber (v: YamlValue) : float =
    match v with
    | VInt i -> float i
    | VFloat f -> f
    | other -> raise (InvalidOperationException(sprintf "expected Int or Float, got %A" other))

let private toTritsList (s: string) : Tri list =
    [ for c in s -> toTri (string c) ]

let private tritsListToStr (ts: Tri list) : string =
    ts |> List.map toStr |> String.concat ""

let private feedbackToStr (fb: Float.FloatFeedback) : string =
    match fb with
    | Float.FloatFeedback.InterpretationSuperposed -> "interpretation-superposed"
    | Float.FloatFeedback.ValueSuperposed -> "value-superposed"

// ---------------------------------------------------------------------------
// Cross-verify fact
// ---------------------------------------------------------------------------

[<Fact>]
let ``cross-verify tri-boolean vectors match TS + F# + C# + Rust`` () =
    let root = repoRoot ()
    let yamlPath = Path.Join(root, "tests", "cross-verification", "tri-boolean", "vectors.yaml")
    let yamlText = File.ReadAllText(yamlPath)

    let doc =
        match parse yamlText with
        | Ok value -> value
        | Error feedback ->
            raise (InvalidOperationException(sprintf "our YAML port declined vectors.yaml: %A" feedback))

    let top = mapEntries doc "<root>"
    let records =
        match field top "vectors" "<root>" with
        | VSeq items -> items
        | other -> raise (InvalidOperationException(sprintf "expected Seq at vectors, got %A" other))

    let results = System.Collections.Generic.Dictionary<string, obj>(StringComparer.Ordinal)

    for i in 0 .. records.Length - 1 do
        let ctx = sprintf "vectors[%d]" i
        let m = mapEntries records.[i] ctx
        let id = asStr (field m "id" ctx) (ctx + ".id")
        let typeStr = asStr (field m "type" ctx) (ctx + ".type")

        if typeStr = "unary" then
            let stateStr = asStr (field m "state" ctx) (ctx + ".state")
            let t = toTri stateStr
            let (measureOk, measureValue, measureFeedback) =
                match measure t with
                | Ok b -> (true, b, "")
                | Error CollapseFeedback.CollapsedLivingUncertainty -> (false, false, "collapsed-living-uncertainty")

            let res = {|
                ``type`` = "unary"
                state = stateStr
                isLiving = isLiving t
                isCertain = isCertain t
                notState = toStr (notTri t)
                cooperateState = toStr (cooperate t)
                measureOk = measureOk
                measureValue = measureValue
                measureFeedback = measureFeedback
                mapNot = toStr (mapTri not t)
                bindNot = toStr (bindTri (fun b -> fromBool (not b)) t)
                bindToT = toStr (bindTri (fun _ -> Tri.T) t)
            |}
            results.[id] <- box res
        else if typeStr = "binary" then
            let leftStr = asStr (field m "left" ctx) (ctx + ".left")
            let rightStr = asStr (field m "right" ctx) (ctx + ".right")
            let left = toTri leftStr
            let right = toTri rightStr

            let res = {|
                ``type`` = "binary"
                left = leftStr
                right = rightStr
                expectedAnd = toStr (andTri left right)
                expectedOr = toStr (orTri left right)
            |}
            results.[id] <- box res
        else if typeStr = "float" then
            let highStr = asStr (field m "high" ctx) (ctx + ".high")
            let decoderStr = asStr (field m "decoder" ctx) (ctx + ".decoder")
            let lowStr = asStr (field m "low" ctx) (ctx + ".low")

            let high = toTritsList highStr
            let decoderVec = toTritsList decoderStr
            let low = toTritsList lowStr

            let f = Float.fromTrits high decoderVec low
            let dRes = Float.decode f

            let (expectedOk, expectedValue, expectedFeedback) =
                match dRes with
                | Ok v -> (true, v, "")
                | Error fb -> (false, 0.0, feedbackToStr fb)

            let res = System.Collections.Generic.Dictionary<string, obj>(StringComparer.Ordinal)
            res.["type"] <- "float"
            res.["high"] <- highStr
            res.["decoder"] <- decoderStr
            res.["low"] <- lowStr
            res.["expectedOk"] <- expectedOk
            res.["expectedValue"] <- expectedValue
            res.["expectedFeedback"] <- expectedFeedback

            match tryField m "encode_value" with
            | Some encodeValNode ->
                let encodeValue = asNumber encodeValNode
                let encRes = Float.fromValue encodeValue f.Shape
                res.["encodeValue"] <- encodeValue
                match encRes with
                | Ok encFloat ->
                    res.["expectedEncodeOk"] <- true
                    res.["expectedEncodeHigh"] <- tritsListToStr encFloat.High
                    res.["expectedEncodeDecoder"] <- tritsListToStr encFloat.Decoder
                    res.["expectedEncodeLow"] <- tritsListToStr encFloat.Low
                | Error detail ->
                    res.["expectedEncodeOk"] <- false
                    res.["expectedEncodeDetail"] <- detail
            | None -> ()

            results.[id] <- box res

    // Write fsharp-output.json
    let options = JsonSerializerOptions(WriteIndented = true)
    let json = JsonSerializer.Serialize(results, options).Replace("\r\n", "\n")
    let outputPath = Path.Join(root, "tests", "cross-verification", "tri-boolean", "fsharp-output.json")
    File.WriteAllText(outputPath, json)

File: tests/Tests.CSharp/TriBoolean/CrossVerifyTests.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;
using Xunit;
using Zeta.Core.CSharp.TriBoolean;
using Zeta.Core.CSharp.Yaml;

namespace Zeta.Tests.CSharp.TriBoolean;

/// <summary>
/// Reads tests/cross-verification/tri-boolean/vectors.yaml via our YAML port,
/// evaluates TriBoolean operations on each vector, and writes cs-output.json.
/// </summary>
public class CrossVerifyTests
{
    private static readonly JsonSerializerOptions JsonOptions = new() { WriteIndented = true };

    private static IReadOnlyList<KeyValuePair<string, YamlValue>> MapEntries(YamlValue v, string ctx) =>
        v is YamlValue.YMap m
            ? m.Entries
            : throw new InvalidOperationException($"expected Map at {ctx}, got {v.GetType().Name}");

    private static YamlValue Field(IReadOnlyList<KeyValuePair<string, YamlValue>> entries, string key, string ctx)
    {
        foreach (var kvp in entries)
            if (string.Equals(kvp.Key, key, StringComparison.Ordinal))
                return kvp.Value;
        throw new InvalidOperationException($"missing field '{key}' at {ctx}");
    }

    private static string AsStr(YamlValue v, string ctx) =>
        v is YamlValue.YStr s
            ? s.Value
            : throw new InvalidOperationException($"expected Str at {ctx}, got {v.GetType().Name}");

    private static string RepoRoot()
    {
        var dir = new DirectoryInfo(Path.GetDirectoryName(typeof(CrossVerifyTests).Assembly.Location)!);
        while (dir is not null && !File.Exists(Path.Join(dir.FullName, "Zeta.sln")))
            dir = dir.Parent;
        return dir?.FullName
            ?? throw new InvalidOperationException("Could not locate repo root (Zeta.sln) from test assembly location.");
    }

    private static Tri ToTri(string s) => s switch
    {
        "T" => Tri.T,
        "F" => Tri.F,
        "N" => Tri.N,
        _ => throw new InvalidOperationException($"unknown tri state: {s}")
    };

    private static string ToStr(Tri t) => t switch
    {
        Tri.TrueCell => "T",
        Tri.FalseCell => "F",
        Tri.NCell => "N",
        _ => throw new InvalidOperationException("unknown cell subclass")
    };

    private static Dictionary<string, object> ProcessUnary(YamlValue val, string ctx)
    {
        var m = MapEntries(val, ctx);
        var stateStr = AsStr(Field(m, "state", ctx), $"{ctx}.state");
        var t = ToTri(stateStr);

        var mRes = TriOps.Measure(t);
        bool measureOk = mRes is MeasureResult.Resolved;
        bool measureValue = mRes is MeasureResult.Resolved r && r.Value;
        string measureFeedback = mRes is MeasureResult.Collapsed c
            ? (c.Feedback == CollapseFeedback.CollapsedLivingUncertainty ? "collapsed-living-uncertainty" : "")
            : "";

        return new Dictionary<string, object>(StringComparer.Ordinal)
        {
            ["type"] = "unary",
            ["state"] = stateStr,
            ["isLiving"] = TriOps.IsLiving(t),
            ["isCertain"] = TriOps.IsCertain(t),
            ["notState"] = ToStr(TriOps.NotTri(t)),
            ["cooperateState"] = ToStr(TriOps.Cooperate(t)),
            ["measureOk"] = measureOk,
            ["measureValue"] = measureValue,
            ["measureFeedback"] = measureFeedback,
            ["mapNot"] = ToStr(TriOps.MapTri(t, b => !b)),
            ["bindNot"] = ToStr(TriOps.BindTri(t, b => TriOps.FromBool(!b))),
            ["bindToT"] = ToStr(TriOps.BindTri(t, _ => Tri.T))
        };
    }

    private static Dictionary<string, object> ProcessBinary(YamlValue val, string ctx)
    {
        var m = MapEntries(val, ctx);
        var leftStr = AsStr(Field(m, "left", ctx), $"{ctx}.left");
        var rightStr = AsStr(Field(m, "right", ctx), $"{ctx}.right");
        var left = ToTri(leftStr);
        var right = ToTri(rightStr);

        return new Dictionary<string, object>(StringComparer.Ordinal)
        {
            ["type"] = "binary",
            ["left"] = leftStr,
            ["right"] = rightStr,
            ["expectedAnd"] = ToStr(TriOps.AndTri(left, right)),
            ["expectedOr"] = ToStr(TriOps.OrTri(left, right))
        };
    }

    [Fact]
    public void CrossVerifyTriBooleanVectorsMatchParity()
    {
        var root = RepoRoot();
        var yamlPath = Path.Join(root, "tests", "cross-verification", "tri-boolean", "vectors.yaml");
        var yamlText = File.ReadAllText(yamlPath);

        ParseResult parsed = YamlDom.Parse(yamlText);
        if (!parsed.Ok)
            throw new InvalidOperationException($"YAML port declined vectors.yaml: {parsed.Feedback}");

        var top = MapEntries(parsed.Value!, "<root>");
        if (Field(top, "vectors", "<root>") is not YamlValue.YSeq seq)
            throw new InvalidOperationException("expected Seq at vectors");

        var results = new Dictionary<string, object>(StringComparer.Ordinal);

        for (int i = 0; i < seq.Items.Count; i++)
        {
            var ctx = $"vectors[{i}]";
            var item = seq.Items[i];
            var m = MapEntries(item, ctx);

            var id = AsStr(Field(m, "id", ctx), $"{ctx}.id");
            var type = AsStr(Field(m, "type", ctx), $"{ctx}.type");

            if (string.Equals(type, "unary", StringComparison.Ordinal))
            {
                results[id] = ProcessUnary(item, ctx);
            }
            else if (string.Equals(type, "binary", StringComparison.Ordinal))
            {
                results[id] = ProcessBinary(item, ctx);
            }
        }

        // Write cs-output.json for compare.ts.
        var outputPath = Path.Join(root, "tests", "cross-verification", "tri-boolean", "cs-output.json");
        var json = JsonSerializer.Serialize(results, JsonOptions);
        var lfJson = json.Replace("\r\n", "\n").Replace("\r", "\n");
        File.WriteAllText(outputPath, lfJson, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
    }
}
